### PR TITLE
8333644: C2: assert(is_Bool()) failed: invalid node class: Phi

### PR DESCRIPTION
**Pull request by chhagedorn**
In the patch of [JDK-8330386](https://bugs.openjdk.org/browse/JDK-8330386), I've added a test case that took the code path in `clone_loop_handle_data_uses()` with the new `OpaqueInitializedAssertionPredicateNode`. So, it proved that we should also add a case for `OpaqueInitializedAssertionPredicate` - yet somehow, I forgot to do that. Unfortunately, I could not come up with a failing test if we do not handle this case separately and the mistake went unnoticed.

But the fuzzer has now found such a case where we crash when creating the Mach graph because we have an `If` with a `Phi` input instead of a `Bool`. This happens exactly because of not handling `OpaqueInitializedAssertionPredicate` in  `clone_loop_handle_data_uses()`. 

We have an outside use of a `4843 Bool` (`4193 RangeCheck`) which is also an input into a `OpaqueInitializedAssertionPredicate` node:

image

We should now clone this `Bool/Cmp` down to avoid having a `Phi` between the `OpaqueInitializedAssertionPredicate` and
the `Bool` node which currently wrongly happens:

image

and we crash later. With this simple patch, this is being avoided.

The good thing is that we now have a test that makes sure that this new condition is properly tested.

Thanks,
Christian

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333644](https://bugs.openjdk.org/browse/JDK-8333644): C2: assert(is_Bool()) failed: invalid node class: Phi (**Bug** - P3)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [68215366](https://git.openjdk.org/jdk/pull/19561/files/682153667fa45f29e7d7ede7f3518fa51bd9eb3c)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19561/head:pull/19561` \
`$ git checkout pull/19561`

Update a local copy of the PR: \
`$ git checkout pull/19561` \
`$ git pull https://git.openjdk.org/jdk.git pull/19561/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19561`

View PR using the GUI difftool: \
`$ git pr show -t 19561`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19561.diff">https://git.openjdk.org/jdk/pull/19561.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19561#issuecomment-2149975813)

### CODE REVIEW

File: src/hotspot/share/opto/loopopts.cpp [hunk 1]
```diff
@@ -2154,18 +2154,19 @@ void PhaseIdealLoop::clone_loop_handle_data_uses(Node* old, Node_List &old_new,
...
-      // Loop predicate If node connects to Bool node through Opaque1 node.
       //
-      // If the use is an AllocateArray through its ValidLengthTest input,
-      // make sure the Bool/Cmp input is cloned down to avoid a Phi between
-      // the AllocateArray node and its ValidLengthTest input that could cause
+      // For several uses we need to make sure that there is no phi between,
+      // the use and the Bool/Cmp. We therefore clone the Bool/Cmp down here
+      // to avoid such a phi in between.
+      // For example, it is unexpected that there is a Phi between an
+      // AllocateArray node and its ValidLengthTest input that could cause
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopopts.cpp:2167]: Was outdated (we already removed `Opaque1` nodes for predicates some time ago) and we also handle more nodes now and not only `AllocateArray`. Improved that a little bit.

File: test/hotspot/jtreg/compiler/predicates/assertion/TestOpaqueInitializedAssertionPredicateNode.java [hunk 2]
```diff
@@ -54,6 +54,17 @@
  * @run main compiler.predicates.assertion.TestOpaqueInitializedAssertionPredicateNode
  */
 
+/*
+ * @test id=clone_loop_handle_data_uses
+ * @bug 8333644
+ * @modules java.base/jdk.internal.misc:+open
+ * @summary Test that using OpaqueInitializedAssertionPredicate for Initialized Assertion Predicates instead of Opaque4
+ *          nodes also work with clone_loop_handle_data_uses() which missed a case before.
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/predicates/assertion/TestOpaqueInitializedAssertionPredicateNode.java:62]:
```suggestion
 *          nodes also works with clone_loop_handle_data_uses() which missed a case before.
```

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Comment by chhagedorn** [comment]
Thanks Tobias for your review!

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by chhagedorn** [comment]
Thanks Vladimir for your review!

/integrate

**Pull request closed without merging**